### ISSUE TITLE
[NCO Bug] Remove misleading "No such file or directory" syntax errors from output files

### ISSUE THREAD

**Issue by KateFriedman-NOAA** [NOAA-EMC/global-workflow#1252]
Bugzilla [#1369](http://www2.spa.ncep.noaa.gov/bugzilla/show_bug.cgi?id=1369)
Bugzilla [#1016](http://www2.spa.ncep.noaa.gov/bugzilla/show_bug.cgi?id=1016)

Details from bugzilla:

From NCO SPA:
```
In the output logs for gfs/gdas/enkfgdas there are hundres of instances of "No such file or directory" messages 
printed out across many different tasks, which makes it more difficult to troubleshoot production failures.
Please work towards removing all related syntax errors from the output logs

A list of tasks from the 12z cycle found with these messages:
enkfgdas_diag_12
enkfgdas_fcst_??_12
gdas_atmos_analysis_diag_12
gdas_atmos_tropcy_qc_reloc_12
gdas_atmos_verfozn_12
gdas_atmos_verfrad_12
gdas_forecast_12
gfs_atmos_gempak_meta_12
gfs_atmos_tropcy_qc_reloc_12
gfs_atmos_wafs_grib2_0p25_12
gfs_atmos_wafs_grib2_12
gfs_mos_ext_grd_prdgen_12
gfs_mos_ext_grd_prep_12
gfs_mos_ext_stn_fcst_12
gfs_mos_ext_stn_prdgen_12
gfs_mos_ext_stn_prep_12
gfs_mos_grd_fcst_12
gfs_mos_grd_prdgen_12
gfs_mos_stn_fcst_12
gfs_mos_stn_prdgen_12
gfs_mos_wx_ext_prdgen_12
gfs_mos_wx_prdgen_12
gfs_wave_prdgen_gridded_12
```

Additional jobs from the GEFS v12 package:
```
gefs_atmos_prdgen_gfs_00.o37433087:rm: cannot remove 'exlist': No such file or directory
gefs_forecast_0_16d_c00_00.o37433286:ls: cannot access /gpfs/dell4/nco/ops/nwges/para/gefs.20200916/00/init/c00/RESTART/coupler: No such file or directory
gefs_forecast_0_16d_c00_00.o37433286:ls: cannot access /gpfs/dell4/nco/ops/com/gefs/para/gefs.20200916/00/wave/rundata/rmp_src_to_dst_conserv_*: No such file
or directory
gefs_wave_prep_c00_00.o37432963:mv: cannot stat 'ice.out': No such file or directory
```

Comment from @RussTreadon-NOAA:
```
A check of the gdas_atmos_analysis_diag and enkfgdas_diag "No such file" messages shows that the messages 
originate from the same script, exglobal_diag.sh.  The following loop is responsible for generating the 
"No such file" messages

   # Restrict diagnostic files containing rstprod data
   rlist="conv_gps conv_ps conv_pw conv_q conv_sst conv_t conv_uv saphir"
   for rtype in $rlist; do
       ${CHGRP_CMD} *${rtype}*
   done

A check on file existence can be added to ensure ${CHGRP_CMD} is only executed when ${rtype} exists.

For GFS v16.3 exgdas_diag.sh resides in NOAA-EMC/GSI branch release/gfsda.v16.3.0.  Note that NOAA-EMC/GSI 
develop no longer maintains DA jobs or scripts.  GFS DA jobs and scripts are now maintained in the 
global-workflow repository.  Therefore, add Rahul Mahajan to the cc list for this bugzilla.
```

Comment from Steven Earle:
```
gfs_mos tasks are excluded from this... but please do a thorough review of all gfs/gdas/enkf tasks for 
version 17... we need to work toward remove all false/misleading errors from our output.
```

**Comment by HenryRWinterbottom** [comment]
@RussTreadon-NOAA I created a branch to address this issue. It can be found [here](https://github.com/HenryWinterbottom-NOAA/global-workflow/tree/feature/gfsv17_issue_1252). Would you please be able to test this branch and confirm whether the above issue is resolved?

If you do, thank you. And if so, can you please pass me the log files if the issue remains such that I can dig in further.

**Comment by HenryRWinterbottom** [comment]
@KateFriedman-NOAA I am seeing the following in the logs when addressing this issue, coming from `ush/jjob_header.sh`:

```
++ jjob_header.sh[81]: setpdy.sh
sed: can't read /scratch1/NCEPDEV/da/Henry.Winterbottom/work/global-workflow/COMROOT/date/t00z: No such file or directory
 completed cleanly
 completed cleanly
Source PDY script to export PDYm7, ..., PDY, ..., PDYp7 variables.
```

Do these fall into the category of problematic messages?

**Comment by WalterKolczynski-NOAA** [comment]
> @KateFriedman-NOAA I am seeing the following in the logs when addressing this issue, coming from `ush/jjob_header.sh`:
> 
> ```
> ++ jjob_header.sh[81]: setpdy.sh
> sed: can't read /scratch1/NCEPDEV/da/Henry.Winterbottom/work/global-workflow/COMROOT/date/t00z: No such file or directory
>  completed cleanly
>  completed cleanly
> Source PDY script to export PDYm7, ..., PDY, ..., PDYp7 variables.
> ```
> 
> Do these fall into the category of problematic messages?

Yes, BUT I think this is a bug with `setpdy.sh` when used on RDHPCS or outside of ops, not anything wrong in workflow. Skip it for now.

**Comment by KateFriedman-NOAA** [comment]
> a bug with setpdy.sh when used on RDHPCS or outside of ops, not anything wrong in workflow

I second @WalterKolczynski-NOAA's reply.

**Comment by HenryRWinterbottom** [comment]
@KateFriedman-NOAA @WalterKolczynski-NOAA 

There are also some `No such file or directory` messages being raised by `/scratch1/NCEPDEV/global/glopara/git/obsproc/v1.1.2/scripts/exglobal_makeprepbufr.sh`. Should these be addressed? If so, not in this PR but one for `obsproc`?

**Comment by KateFriedman-NOAA** [comment]
Yeah, that's most likely an obsproc thing to resolve on their end. Can you share a sample log so we can see the particular messages though? Thanks!

**Comment by HenryRWinterbottom** [comment]
Please see `/scratch1/NCEPDEV/da/Henry.Winterbottom/work/global-workflow/COMROOT/x001_gfsv17_issue_1252/logs/2021122100/gdasprep.log`. 

The `stmp` path is gone, but grep for `No such file or directory` and you should see what I referenced above.

**Comment by HenryRWinterbottom** [comment]
There are also similar messages coming from the TC tracker and genesis applications. See respectively the following:

`/scratch1/NCEPDEV/da/Henry.Winterbottom/work/global-workflow/COMROOT/x001_gfsv17_issue_1252/logs/2021122100/gfstracker.log`

`/scratch1/NCEPDEV/da/Henry.Winterbottom/work/global-workflow/COMROOT/x001_gfsv17_issue_1252/logs/2021122100/gfsgenesis.log`

For the CI `C96_atm3DVar`, those are the remaining `No such file or directory` instances.

**Comment by WalterKolczynski-NOAA** [comment]
I think the pdy message is in every job we run.

**Comment by HenryRWinterbottom** [comment]
> I think the pdy message is in every job we run.

Yes, so far it is showing up in all of the log files for the experiment referenced [above](https://github.com/NOAA-EMC/global-workflow/issues/1252#issuecomment-2152997851).

@WalterKolczynski-NOAA If this is worth addressing, please open a new issue and I will self assign and fix it, if possible.

**Comment by KateFriedman-NOAA** [comment]
> Please see `/scratch1/NCEPDEV/da/Henry.Winterbottom/work/global-workflow/COMROOT/x001_gfsv17_issue_1252/logs/2021122100/gdasprep.log`.

Those "No such"s are almost all from being unable to remove (because the file didn't already exist, which is normal) or are a file we don't need/use. The obsproc group needs to improve the error handling to remove these messages in their scripts. For this issue, I would ignore those "No such"s in the prep job.

**Comment by HenryRWinterbottom** [comment]
> > Please see `/scratch1/NCEPDEV/da/Henry.Winterbottom/work/global-workflow/COMROOT/x001_gfsv17_issue_1252/logs/2021122100/gdasprep.log`.
> 
> Those "No such"s are almost all from being unable to remove (because the file didn't already exist, which is normal) or are a file we don't need/use. The obsproc group needs to improve the error handling to remove these messages in their scripts. For this issue, I would ignore those "No such"s in the prep job.

Thank you @KateFriedman-NOAA for checking my work. I just wanted to make sure this wasn't our responsibility or a fix required of this bugzilla/PR.

**Comment by KateFriedman-NOAA** [comment]
> `/scratch1/NCEPDEV/da/Henry.Winterbottom/work/global-workflow/COMROOT/x001_gfsv17_issue_1252/logs/2021122100/gfstracker.log`
> 
> `/scratch1/NCEPDEV/da/Henry.Winterbottom/work/global-workflow/COMROOT/x001_gfsv17_issue_1252/logs/2021122100/gfsgenesis.log`

Pretty much the same thing as the prep job...they are "cannot remove" "No such"s. If they are in scripts that global-workflow owns then we should improve the error handling to not complain if we can't remove a file that doesn't exist (i.e. check if the file exists first and then remove if it does, otherwise nothing). If these are from scripts that the TC_tracker repo owns then we should work with them to clean up.

**Comment by DavidHuber-NOAA** [comment]
Opened https://github.com/NOAA-EMC/TC_tracker/issues/8 to address issues in the TC Tracker.

**Comment by DavidHuber-NOAA** [comment]
Opened https://github.com/NOAA-EMC/obsproc/issues/85 to address issues in obsproc.  Note that this issue is not strictly an NCO issue since these scripts are not run as part of the operational workflow.

**Comment by KateFriedman-NOAA** [comment]
Thanks @DavidHuber-NOAA !

**Comment by KateFriedman-NOAA** [comment]
The list provided at the start of this issue should now be (after removing jobs that aren't in the system or don't need to be reviewed as part of this based on NCO feedback in the bugzilla):
```
enkfgdas_diag_12
enkfgdas_fcst_??_12
gdas_atmos_analysis_diag_12
gdas_atmos_tropcy_qc_reloc_12
gdas_atmos_verfozn_12
gdas_atmos_verfrad_12
gdas_forecast_12
gfs_atmos_gempak_meta_12
gfs_atmos_tropcy_qc_reloc_12
gfs_wave_prdgen_gridded_12
```

Review all jobs but double check ^ these jobs in particular.

**Comment by KateFriedman-NOAA** [comment]
The list provided at the start of this issue should now be (after removing jobs that aren't in the system or don't need to be reviewed as part of this based on NCO feedback in the bugzilla):
```
enkfgdas_diag_12
enkfgdas_fcst_??_12
gdas_atmos_analysis_diag_12
gdas_atmos_tropcy_qc_reloc_12
gdas_atmos_verfozn_12
gdas_atmos_verfrad_12
gdas_forecast_12
gfs_atmos_gempak_meta_12
gfs_atmos_tropcy_qc_reloc_12
gfs_wave_prdgen_gridded_12
```

Review all jobs but double check ^ these jobs in particular.

**Comment by KateFriedman-NOAA** [comment]
Ran the full suite of CI tests that can run on Hera currently and examined logs for "No such" statements. The only "No such" statements observed were from the following sources:

1. an operational (WCOSS2-only) date utility that doesn't work on the R&Ds but isn't an actual issue (e.g. `sed: can't read /scratch1/NCEPDEV/stmp4/Kate.Friedman/RUNTESTS_test/COMROOT/date/t06z: No such file or directory`)
2. the obsproc/prepobs packages (outside of global-workflow control but a separate issue has been opened with those repos)
3. the tracker/genesis package (outside of global-workflow control but reported to tracker team)

No other "No such" instances were observed in the logs for the CI tests on Hera.

The issue reported in the `exgdas_diag.sh` script (now `exglobal_diag.sh`) was resolved via PR #2688.

**Comment by KateFriedman-NOAA** [comment]
Will examine all `rm`, `ls`, and `find` instances to ensure correct flags and existence checks exist.

**Comment by KateFriedman-NOAA** [comment]
> A check of the gdas_atmos_analysis_diag and enkfgdas_diag "No such file" messages shows that the messages 
originate from the same script, exglobal_diag.sh.  The following loop is responsible for generating the 
"No such file" messages
>```
>   # Restrict diagnostic files containing rstprod data
>   rlist="conv_gps conv_ps conv_pw conv_q conv_sst conv_t conv_uv saphir"
>   for rtype in $rlist; do
>       ${CHGRP_CMD} *${rtype}*
>   done
>```
>A check on file existence can be added to ensure ${CHGRP_CMD} is only executed when ${rtype} exists.

A check of `scripts/exglobal_diag.sh` shows the issue noted in the original issue comment has been resolved via a check for `rfile` "exists and has a size greater than zero". See line 245:
```
241    # Restrict diagnostic files containing rstprod data
242    rlist="conv_gps conv_ps conv_pw conv_q conv_sst conv_t conv_uv saphir"
243    for rtype in $rlist; do
244       for rfile in *"${rtype}"*; do
245          if [[ -s "${rfile}" ]]; then
246             ${CHGRP_CMD} "${rfile}"
247          fi
248       done
249    done
```

**Comment by KateFriedman-NOAA** [comment]
Known "No such" statements to address outside of this issue:

1. `PDY` from `jjob_header`: `../ush/jjob_header.sh: line 87: ./PDY: No such file or directory`
2. numerous "No such" statements from external packages: prep and tracker/genesis

The list from the original issue comment but now without the jobs that do not live in GFSv17 anymore (wafs and mos):
```
enkfgdas_diag_12
enkfgdas_fcst_??_12
gdas_atmos_analysis_diag_12
gdas_atmos_tropcy_qc_reloc_12
gdas_atmos_verfozn_12
gdas_atmos_verfrad_12
gdas_forecast_12
gfs_atmos_gempak_meta_12
gfs_atmos_tropcy_qc_reloc_12
gfs_wave_prdgen_gridded_12
```
### enkfgdas_fcst_??_12 & gdas_forecast_12

Forecast job scripts have been rewritten and now use file copying utilities that announce if a file is missing without false positives.

### enkfgdas_diag_12 & gdas_atmos_analysis_diag_12

Examined job and made change to remove unnecessary `$mkdata` check and removal of `$DATA` within ex-script. All `$DATA` removal should be done at end of JJOB script. No other "No such" messages observed from running these jobs that run `scripts/exglobal_diag.sh`.

### gdas_atmos_tropcy_qc_reloc_12 & gfs_atmos_tropcy_qc_reloc_12

This job only runs in ops or during real-time parallels so unable to fully test in retrospective dev mode. The `break` commands will be removed from the ex-script to reduce the errors they throw.

### gdas_atmos_verfozn_12 & gdas_atmos_verfrad_12

No errors observed in tests of job on WCOSS2 and only shows the known `PDY` message (see above).

### gfs_atmos_gempak_meta_12

No errors observed in tests of job on WCOSS2 and only shows the known `PDY` message (see above).

### gfs_wave_prdgen_gridded_12

Job has some other issues to resolve beyond the "No such" statements. Working to resolve.

**Comment by aerorahul** [comment]
@DavidHuber-NOAA @KateFriedman-NOAA Thanks for resolving this long standing nco-bug.
What do you think of adding a step (not a job) in our CI system that scraps for strings such as "No such file or directory", "WARNING", "FATAL ERROR", so that they are caught early in the development cycle?  We could open an `[Enabler]` issue to that end.

**Comment by DavidHuber-NOAA** [comment]
I like that idea!  I have a script that I run offline to that end and screens out warnings, etc, that are known to exist, but it could use some beautifying.  I could throw it at the enabler issue.

**Comment by KateFriedman-NOAA** [comment]
@aerorahul I like that idea. I would also include "unbound variable" warnings.
